Tooltips and Backing Dialog

monotkate opened the pull request:
Added Tooltips.  Opening for discussion some of the wording and appearance.  I worded backing tooltips as contributions to potentially get people used to the term back, but can change it easily.  

Backing now triggers a dialog window.  This window can be triggered from either the little person icon next to the like (feel free to suggest a different icon), or from the backers window with the button.  If you have already created a back, the little person will trigger an edit instead of an add (tooltip updates as appropriate).  You can also edit from the backers page by clicking on the edit next to your back.  Which I just realized could use a tooltip.  If the window is clicked off in the middle of an edit, it will not submit the edit.  

This is going to fail one of the syntax tests because it's using the dialogControllerBack before it's defined.  I would rather leave the definition of it at the bottom instead of the top, however, because it would be obnoxious at the top.  Is there a way to declare ahead of time and define later or am I thinking of another language.  Having trouble finding it tonight.

Other than that, code should be commented, pass the other checks, and pass tests.  I didn't update the regular tests because while it functions differently on the front end, most all of the actual addition and edit submission is identical.  I can add if necessary though.

There may be a few quirks left that I can go back and fix.  I wanted to get some opinions on the appearance of dialog/graphics/tooltips/wording for now, and the backing dialog is functional.  There are certain things I'm not absolutely thrilled with, but they work, and I'm trying to gauge if it needs to be changed or not.  Sometimes a difficult assessment when you're trying to teach people Logix programming at the same time.

I'm tired and braindead for now.  Please discuss (tear apart) what's up there so I can change it as necessary for a better user experience.  I'm noticing some other things that I'd like to tweak as well, but I might leave that separate from the backers dialog pull.

@YashdalfTheGray @alanlgirard @Mctalian @Daviddt17 @LongArmMcGee @bcbrennecke @davethenipper

YashdalfTheGray commented:
Your travis build failed.

YashdalfTheGray commented on src/assets/app.css:
```diff
@@ -193,3 +209,7 @@ body {
     width: 100%;
 }
+
+.backBox {
```
change this class to `back-box`. CSS classes aren't named in `camelCase`, they instead use `kebab-case`.

YashdalfTheGray commented on src/ideas/ideaBack/ideaAddBack.tpl.html:
```diff
@@ -0,0 +1,40 @@
+<md-dialog class="backBox" aria-label="BackIdea" ng-cloak>
```
Aria labels should be written in normal english since it will be used by accessibility services to pronounce what the user is pointing at. Computer readers will assume `backIdea` is one word instead of being smart enough to know that they are two words.

monotkate commented:
I know. I mentioned it would

YashdalfTheGray commented:
Oh fail, sorry. I think you're going to have to move it to the top, it's not syntactically incorrect but it's better for readability.

YashdalfTheGray commented on src/ideas/ideasView/ideasView.js:
```diff
@@ -435,3 +502,74 @@ angular.module('flintAndSteel')
+
+// Dialog Controller used for controlling the behavior of the dialog
+//   used for login.
+function DialogControllerBack($scope, $mdDialog, ideaSvc, backingObj) {
```
Angular controllers in our project are named like viewName + "Ctrl" so for this, I would change it to `BackDialogCtrl`.

monotkate commented:
Damn. Alright, I'll move it tomorrow. I was hoping there was something nicer.

monotkate commented:
Just realized I think I forgot to actually carry out the purpose of my card. Go me. I'll work on that yet too. Should be an easy fix.

YashdalfTheGray commented on src/ideas/ideasView/ideasView.tpl.html:
```diff
@@ -54,20 +57,46 @@
+            <span flex ng-if="!hasUserBacked()">
+                <md-button class="md-icon-button" aria-label="Back" ng-click="showAddBack($event)">
+                    <i class="material-icons icon-button-adjust">&#xE7FE;</i>
+                    <md-tooltip md-direction="bottom">
```
You can move the `ng-if="!hasUserBacked()"` down to the tooltip declaration or even further by wrapping the "Contribute" in a `<div>` and putting the `ng-if` on that.

```html
<md-tooltip md-direction="bottom">
    <div ng-if="hasUserBacked()">Edit Contribution</div>
    <div ng-if="!hasUserBacked()">Contribute</div>
</md-tooltip>
```

monotkate commented:
Woo!  It should now require both at least one backing type and a comment in order to submit an edit or add back.  Also fixed the other issues you pointed out @YashdalfTheGray and moved the Dialog code to the top so it passes build

Mctalian commented:
Sounds like good stuff. I'll try to get to this before peer programming
tonight. If not, a bunch of us will look it over there.
On Dec 3, 2015 10:01 AM, "monotkate" <notifications@github.com> wrote:

> Woo! It should now require both at least one backing type and a comment in
> order to submit an edit or add back. Also fixed the other issues you
> pointed out @YashdalfTheGray <https://github.com/YashdalfTheGray> and
> moved the Dialog code to the top so it passes build
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-161666050>
> .
>

Mctalian commented on src/ideas/ideaBack/ideaAddBack.tpl.html:
```diff
@@ -0,0 +1,43 @@
+        </md-toolbar>
+        <md-dialog-content>
+            <!-- Personal Contribution -->
+            <md-input-container flex>
```
Can we put this below the Types of backing and have it be an optional "More Details" or "Anything else?" Text field? (We'll have to limit the number of characters at some point, like the rest of the forms on the site).

We keep having discussions of "What is the user actually expected to type here?" We think putting this below the backing types, changing the "help text," and making it optional would solve that problem.

monotkate replied:
I was wondering if it should be more of a "describe the skills you'd like to add" type thing.  The tagging is a nice way to generalize skills, but doesn't give a lot of detail into the skills that people actually bring to the table.  I can move it and make it optional for now, but I guess this should be a discussion point.

Mctalian commented on src/ideas/ideaBack/ideaAddBack.tpl.html:
```diff
@@ -0,0 +1,43 @@
+                <p>Tag the type of your contribution</p>
+                <div layout="row" layout-wrap>
+                    <div ng-repeat="type in types" flex="50" ng-click="toggle(type, $index)">
+                        <md-checkbox>
```
There is a blue outline around these items when you select them. It's probably a :focus or :selected or :active pseudo class. I think removing the outline is the best way to go. Alan agrees.

Mctalian commented on src/ideas/ideaBack/ideaEditBack.tpl.html:
```diff
@@ -0,0 +1,42 @@
+<md-dialog class="back-box" aria-label="edit idea">
```
Everything in the above file applies here as well.

Mctalian commented on src/ideas/ideaItem/ideaItem.tpl.html:
```diff
@@ -4,10 +4,10 @@
         <div layout="row">
             <div layout="row" layout-align="center center">
-                <identicon username='idea.author.username' size='18'></identicon>
+                <identicon username='idea.author.username' size='14'></identicon>
```
This size change makes the identicon get offset to the upper left but that's a 3rd party app issue. Nothing for you to do here.

monotkate replied:
yeah, I saw it shifts it a bit, but at least personally I hated the author text being the same size as the ideas.  It doesn't make sense imo

Mctalian replied:
I agree. I also just fixed it sort of. The smaller the identicon the harder it is to not have the shift. Is "15" an awful compromise? If/when the fix goes through to angular-identicon, "15" would work nicely.

YashdalfTheGray replied:
I like to stick to even numbers since resolutions are usually even numbers and all the paddings and spacing Material recommends are even but if that really does work well, I could handle 15 too.

Mctalian replied:
It could work, I need to adjust the calculations for small identicons but
they are a pain in the ass. Why can't the world be symmetrical and perfect!?
On Dec 5, 2015 11:50 AM, "Yash" <notifications@github.com> wrote:

> In src/ideas/ideaItem/ideaItem.tpl.html
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#discussion_r46760269>
> :
>
> > @@ -4,10 +4,10 @@
> >          <div><strong>{{idea.title}}</strong></div>
> >          <div layout="row">
> >              <div layout="row" layout-align="center center">
> > -                <identicon username='idea.author.username' size='18'></identicon>
> > +                <identicon username='idea.author.username' size='14'></identicon>
>
> I like to stick to even numbers since resolutions are usually even numbers
> and all the paddings and spacing Material recommends are even but if that
> really does work well, I could handle 15 too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188/files#r46760269>
> .
>

Mctalian commented on src/ideas/ideasView/ideasView.tpl.html:
```diff
@@ -54,20 +57,37 @@
-                <i class="material-icons icon-button-adjust">&#xE87D;</i>
-            </md-button>
-        </span>
         <span flex>{{idea.likes.length}} <a class="idea-likes-click" ng-click="openLikes($event, idea.likes)">likes</a> and {{idea.backs.length}} backs</span>
```
Can we put the `{{idea.likes.length}}` in the anchor tag so both the "likes" and the number describing it are both clickable.

monotkate replied:
Alright so, I'll push in a sec, but I just nixed likes all together.  It looks cleaner and the tooltips should explain what's going on. It doesn't give you the ability to see who liked the idea, but tbh I didn't know that was even there until you mentioned it.  Not a typical feature. Vote for a better icon for backs is completely up for grabs as well.  Right now I used to group icon.

Mctalian commented on src/ideas/ideasView/ideasView.tpl.html:
```diff
@@ -54,20 +57,37 @@
+                    <i class="material-icons icon-button-adjust">&#xE7FE;</i>
+                </md-button>
+                <md-tooltip ng-if="!hasUserBacked()" md-direction="bottom">Contribute</md-tooltip>
+                <md-tooltip ng-if="hasUserBacked()" md-direction="bottom">Edit Contribute</md-tooltip>
```
Can we change this text to "Edit Contribution"?

monotkate replied:
So I used contribution in the tooltips because I know you guys said that people had trouble with the word "back" but I was trying to keep most of the actual procedure with the backing as backs.  @YashdalfTheGray may have to weigh in on this one, because I'm hesitant to change too much to contribution.

Mctalian commented on src/ideas/ideasView/ideasView.tpl.html:
```diff
@@ -54,20 +57,37 @@
+            </span>
+            <span flex>
+                <md-button class="md-icon-button" aria-label="Back" ng-click="showAddBack($event)">
+                    <i class="material-icons icon-button-adjust">&#xE7FE;</i>
```
Similar to the Like "heart" getting filled in when someone likes the idea, can we have a different icon symbolizing that they backed the idea? Is there a pencil next to a person icon on Google Icons in addition to the person with the plus sign?

Mctalian replied:
Looking at the icon list, maybe `person` and `person outline`? Kind similar to `favorite` and `favorite outline` for our "Likes."

Mctalian commented on src/ideas/ideasView/ideasView.tpl.html:
```diff
@@ -154,10 +174,13 @@
             <md-tab label="Backers">
                 <md-content class="md-padding">
+                    <md-button class="md-button md-raised md-primary padding-sides-20" ng-if="!hasUserBacked() && isUserLoggedIn()" ng-click="showAddBack($event)">
+                        Contribute to idea
```
Can we have a similar button here for "Edit Contribution" here?

monotkate replied:
I could easily change this, but I was trying to leave editting the same as the other types.  Consistency.  If you think it would help with UI I could add it, but like my comment before I would probably leave it as "edit back" instead of "edit contribution"

Mctalian replied:
I'm fine with the way it is now.

Mctalian commented on src/index.html:
```diff
@@ -63,14 +66,16 @@
                 </md-button>
 
                 <!-- Profile button -->
                 <div ng-show="isUserLoggedIn()" style="font-size:14px">
```
Can we make all of this clickable, not just the chevron?

LongArmMcGee commented:
Omg bob I feel bad for kate after reading your comments

YashdalfTheGray commented:
Miranda?

Mctalian commented:
We discussed the origins of her username during peer programming, so he was
just playing off that story, I think.
On Dec 4, 2015 12:05 AM, "Yash" <notifications@github.com> wrote:

> Miranda?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-161877784>
> .
>

monotkate commented:
haha!  Who knew the origin of my username?? @alanlgirard ?

Mctalian commented:
Yup!

monotkate commented:
Check out the new code and display.  Opinions welcome for backing icon, new likes/backs display, and whether or not we should add an edit button for backs when you can edit at the back or from the little icon.  Also changing it to contribution or not.  @YashdalfTheGray @Mctalian in particular, anyone else feel free to respond.

Mctalian commented:
Got a few meetings to attend the rest of the day. I'm sorry if I don't get to this in a timely manner.

@YashdalfTheGray , can you give @alanlgirard contributor rights so he can pull @monotkate 's changes locally to get a good look at the visuals in case I fall off the face of the earth in my busy-ness? Alan and I seemed to be on the same page, so if he can look at it in my stead if I'm too busy, that'd help move this PR along.

YashdalfTheGray commented on src/ideas/ideaBack/ideaEditBack.tpl.html:
```diff
@@ -0,0 +1,39 @@
+        <md-dialog-content>
+            <!-- Edit your contribution type -->
+            <md-input-container flex>
+                <p>Edit your contribution type (at least one)</p>
```
Maybe say "Select at least one"? I don't know how to phrase that correctly.

Mctalian replied:
"Select at least one" in small, gray accent text would probably be fine.

YashdalfTheGray commented on src/ideas/ideasView/ideasView.js:
```diff
@@ -3,6 +3,77 @@
 
+// Dialog Controller used for controlling the behavior of the dialog
+//   used for login.
+function DialogCtrlBack($scope, $mdDialog, ideaSvc, backingObj) {
```
`DialogBackCtrl`?

YashdalfTheGray commented on src/index.html:
```diff
@@ -55,7 +55,10 @@
-            <h2><span><img ui-sref="home" class="center-img nav-bar-pointer" src="assets/Innovation_Challenge_Bulb.png" width="25"/></span></h2>
+            <h2><span>
+                <img ui-sref="home" class="center-img nav-bar-pointer" src="assets/Innovation_Challenge_Bulb.png" width="25"/>
+                <md-tooltip md-direction="bottom">Home</md-tooltip>
```
Good call on this! :100:

Mctalian commented on src/ideas/ideasView/ideasView.tpl.html:
```diff
@@ -54,20 +57,65 @@
+                    {{idea.backs.length}}
+                    <i class="material-icons icon-button-adjust">&#xE7FB;</i>
+                </md-button>
+                <md-tooltip md-direction="bottom">Edit Contribute</md-tooltip>
```
"Edit Contribute" to me seems like we translated this from another language and it didn't quite translate properly. I'd prefer "Contribution" to make the grammar correct, or "Back." Whichever you prefer.

monotkate replied:
Fixed this.  That was a typo, whoops

Mctalian replied:
Ah ok, at some point I thought I remembered bringing this up and having someone disagree with me, so I was restating my opinion. But apparently I'm imagining things and this was just a typo haha, my bad

Mctalian commented:
Approved. Looks good!

YashdalfTheGray commented:
I prefer back but yeah, definitely "contribution". Also, looks good.


Yash Kulshrestha
yash.kulshrestha@gmail.com
(614)563-1153

On Sat, Dec 5, 2015 at 12:49 PM, Robert <notifications@github.com> wrote:

> Approved. Looks good!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-162234375>
> .
>

monotkate commented:
I put "edit backing" or "add back".  You can debate that among yourselves again.  Also changed the icons for add/edit in the backers window.  There's a delete in there that's commented out while Yash works through the backend stuff.  Remove interaction did not work for backs as far as I could tell working with it yesterday.  Approve again? @Mctalian

monotkate commented:
So I accidentally broke it and then effed up my code.  I think I re-fixed it.  Guys take a look to see if I'm missing something? @YashdalfTheGray @Mctalian

YashdalfTheGray commented:
Approved.

monotkate commented:
@Mctalian did you actually approve this?

Mctalian commented:
Haven't taken a look since the last ask for a review. Will review today.
On Dec 6, 2015 3:09 PM, "monotkate" <notifications@github.com> wrote:

> @Mctalian <https://github.com/Mctalian> did you actually approve this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-162342768>
> .
>

alanlgirard commented on src/ideas/ideasView/ideasView.tpl.html:
```diff
@@ -152,12 +199,29 @@
+                    <!-- Edit current backing -->
+                    <md-button class="md-icon-button" ng-if="isUserLoggedIn()" ng-click="showAddBack($event)" ng-disabled="!hasUserBacked()">
+                        <i class="material-icons icon-button-adjust">&#xE3C9;</i>
+                        <md-tooltip md-direction="bottom">Edit Back</md-tooltip>
```
Should be the same text as the edit icon next to the likes icon.  Should be "Edit Contribution" here or the top icon should be changed to "Edit Back" depending on what wording we want to go with.

alanlgirard commented:
Approved pending small wording change for consistency.

monotkate commented:
@alanlgirard fixed.  I'll merge once it passes the tests officially

Mctalian commented:
@monotkate said:
> There's a delete in there that's commented out while Yash works through the backend stuff. Remove interaction did not work for backs as far as I could tell working with it yesterday.

OH, Ok! `removeinteraction` wasn't going to work all along and that was by design. It was definitely missing the back-end support for it and since I didn't see a card for it at that time, I didn't bother making sure it worked. So the `removeInteraction` call should still be the call to use once we are able to actually delete backs through the back-end. It still might not work, but if you run into issues, please ask me and we can look at it together. I'd like to keep using `removeInteraction` like the rest of the interactions.

(Sorry this was bugging me, so I've been thinking about it since like Monday and wanted to make sure the next step on this was clear)

monotkate commented:
yeah, I was waiting for Yash to finish back-end before delving into this one :+1:

Mctalian commented on src/ideas/ideasView/ideasView.tpl.html:
```diff
@@ -152,12 +199,29 @@
+                        <md-tooltip md-direction="bottom">Edit Contribution</md-tooltip>
+                    </md-button>
+                    <!-- Delete current backing -->
+                    <!--<md-button class="md-icon-button" ng-if="isUserLoggedIn()" ng-click="removeBack()" ng-disabled="!hasUserBacked()">
```
Ah, I see why that wouldn't work here. We don't have a back to grab from. We would need a method that would go through the backs, find the one that is the user's back, and pass that as `back` in `removeInteraction('backs', back)`

I'm actually convinced this would work out of the box. It's just a trickier scenario.

If it was a hovered scenario like comments and updates, it would be much simpler.

Mctalian replied:
@monotkate

monotkate replied:
I have poured over this.  I did discover that part of the issue was that the backers object didn't actually follow the db schema, which of course doesn't help.  My current PR should fix this issue.  However, once it was fixed, I was delivering it exactly what's the in schema and it doesn't work.  I'm not sure if you're in the office next week.  I was going to say if you've got a chance you could come down to my side of the building and attempt to work through it at some point.  I'm stumped.  I don't like to be stumped.

Mctalian replied:
Sure, ping me this week and I can stop by and take a look.

Mctalian replied:
This is likely related to the most recent discussions in #199 but I have not confirmed this. We'll take a closer look later.

monotkate replied:
I figured after seeing that note

Mctalian commented:
@monotkate I lied. Please see my new line comment, I think deleting can be done without any backend changes.

monotkate commented:
haha gotcha.  I'll poke at that once I track down the #199 issue.

Mctalian commented:
:+1:

monotkate commented:
Nope, this is a backend issue.  I'm actually passing it the backing object super easily in the code, it's just not in the function call. 

I tracked it to the backend as well.  Something funky is going on with your new interaction setup.  the db function for updateOnePullArray is getting my back object, but it's not deleting it.  This is the same as the other issue where I just don't know where to troubleshoot it past that point.

![image](https://cloud.githubusercontent.com/assets/11742296/11753430/14bfd3ea-a013-11e5-8693-cf0fc4535b03.png)
![image](https://cloud.githubusercontent.com/assets/11742296/11753453/28cd6faa-a013-11e5-8044-2980e0fd7035.png)

I've never been entirely sure why this one is called updateOnePullArray and then updateOnePushArray either, but here's my console log:
![image](https://cloud.githubusercontent.com/assets/11742296/11753481/4cc7d76a-a013-11e5-8a15-c3643afc390e.png)

Mctalian commented:
One adds an interaction to the array of that particular interaction (push
to the array of backs for example) the other removes a particular
interaction from the appropriate array of interactions (pull from the array
of backs for example). In Mongo, $push and $pull are atomic in that if
someone else calls a $push or $pull to the same idea, or even same
interaction, they do not affect each other.

So in the context here, we're trying to pull a particular element from the
array of backs.

Does this at least shed light on the push and pull?
On Dec 11, 2015 2:28 PM, "monotkate" <notifications@github.com> wrote:

> Nope, this is a backend issue. I'm actually passing it the backing object
> super easily in the code, it's just not in the function call.
>
> I tracked it to the backend as well. Something funky is going on with your
> new interaction setup. the db function for updateOnePullArray is getting my
> back object, but it's not deleting it. This is the same as the other issue
> where I just don't know where to troubleshoot it past that point.
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/11742296/11753430/14bfd3ea-a013-11e5-8693-cf0fc4535b03.png>
> [image: image]
> <https://cloud.githubusercontent.com/assets/11742296/11753453/28cd6faa-a013-11e5-8044-2980e0fd7035.png>
>
> I've never been entirely sure why this one is called updateOnePullArray
> and then updateOnePushArray either, but here's my console log:
> [image: image]
> <https://cloud.githubusercontent.com/assets/11742296/11753481/4cc7d76a-a013-11e5-8a15-c3643afc390e.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-164026695>
> .
>

Mctalian commented:
Ah, we need to remove isInTeam field from the back object! Check out
removeInteraction how we remove the author object and also remove isInTeam
and it should work! :)
On Dec 11, 2015 3:16 PM, "Robert Anderson" <r-anderson.9@onu.edu> wrote:

> One adds an interaction to the array of that particular interaction (push
> to the array of backs for example) the other removes a particular
> interaction from the appropriate array of interactions (pull from the array
> of backs for example). In Mongo, $push and $pull are atomic in that if
> someone else calls a $push or $pull to the same idea, or even same
> interaction, they do not affect each other.
>
> So in the context here, we're trying to pull a particular element from the
> array of backs.
>
> Does this at least shed light on the push and pull?
> On Dec 11, 2015 2:28 PM, "monotkate" <notifications@github.com> wrote:
>
>> Nope, this is a backend issue. I'm actually passing it the backing object
>> super easily in the code, it's just not in the function call.
>>
>> I tracked it to the backend as well. Something funky is going on with
>> your new interaction setup. the db function for updateOnePullArray is
>> getting my back object, but it's not deleting it. This is the same as the
>> other issue where I just don't know where to troubleshoot it past that
>> point.
>>
>> [image: image]
>> <https://cloud.githubusercontent.com/assets/11742296/11753430/14bfd3ea-a013-11e5-8693-cf0fc4535b03.png>
>> [image: image]
>> <https://cloud.githubusercontent.com/assets/11742296/11753453/28cd6faa-a013-11e5-8044-2980e0fd7035.png>
>>
>> I've never been entirely sure why this one is called updateOnePullArray
>> and then updateOnePushArray either, but here's my console log:
>> [image: image]
>> <https://cloud.githubusercontent.com/assets/11742296/11753481/4cc7d76a-a013-11e5-8a15-c3643afc390e.png>
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-164026695>
>> .
>>
>

Mctalian commented:
Explanation: the DB looks for an object that matches the exact query, since
we're passing objects we need to pass the exact object as seen in the DB.
It needs to follow the database schema in other words and isInTeam is not
in the real schema, just something we use on the client side for building
the team up. Those objects aren't guaranteed to have a unique identifier
since they roll up under the idea itself, so we need to search for the
entire object itself. (i.e. comments have an _id, but updates do not)
On Dec 11, 2015 3:17 PM, "Robert Anderson" <r-anderson.9@onu.edu> wrote:

> Ah, we need to remove isInTeam field from the back object! Check out
> removeInteraction how we remove the author object and also remove isInTeam
> and it should work! :)
> On Dec 11, 2015 3:16 PM, "Robert Anderson" <r-anderson.9@onu.edu> wrote:
>
>> One adds an interaction to the array of that particular interaction (push
>> to the array of backs for example) the other removes a particular
>> interaction from the appropriate array of interactions (pull from the array
>> of backs for example). In Mongo, $push and $pull are atomic in that if
>> someone else calls a $push or $pull to the same idea, or even same
>> interaction, they do not affect each other.
>>
>> So in the context here, we're trying to pull a particular element from
>> the array of backs.
>>
>> Does this at least shed light on the push and pull?
>> On Dec 11, 2015 2:28 PM, "monotkate" <notifications@github.com> wrote:
>>
>>> Nope, this is a backend issue. I'm actually passing it the backing
>>> object super easily in the code, it's just not in the function call.
>>>
>>> I tracked it to the backend as well. Something funky is going on with
>>> your new interaction setup. the db function for updateOnePullArray is
>>> getting my back object, but it's not deleting it. This is the same as the
>>> other issue where I just don't know where to troubleshoot it past that
>>> point.
>>>
>>> [image: image]
>>> <https://cloud.githubusercontent.com/assets/11742296/11753430/14bfd3ea-a013-11e5-8693-cf0fc4535b03.png>
>>> [image: image]
>>> <https://cloud.githubusercontent.com/assets/11742296/11753453/28cd6faa-a013-11e5-8044-2980e0fd7035.png>
>>>
>>> I've never been entirely sure why this one is called updateOnePullArray
>>> and then updateOnePushArray either, but here's my console log:
>>> [image: image]
>>> <https://cloud.githubusercontent.com/assets/11742296/11753481/4cc7d76a-a013-11e5-8a15-c3643afc390e.png>
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-164026695>
>>> .
>>>
>>

monotkate commented:
why doesn't it error then if it doesn't work?

On Fri, Dec 11, 2015 at 3:28 PM, Robert <notifications@github.com> wrote:

> Explanation: the DB looks for an object that matches the exact query, since
> we're passing objects we need to pass the exact object as seen in the DB.
> It needs to follow the database schema in other words and isInTeam is not
> in the real schema, just something we use on the client side for building
> the team up. Those objects aren't guaranteed to have a unique identifier
> since they roll up under the idea itself, so we need to search for the
> entire object itself. (i.e. comments have an _id, but updates do not)
>
> On Dec 11, 2015 3:17 PM, "Robert Anderson" <r-anderson.9@onu.edu> wrote:
>
> > Ah, we need to remove isInTeam field from the back object! Check out
> > removeInteraction how we remove the author object and also remove
> isInTeam
> > and it should work! :)
> > On Dec 11, 2015 3:16 PM, "Robert Anderson" <r-anderson.9@onu.edu> wrote:
> >
> >> One adds an interaction to the array of that particular interaction
> (push
> >> to the array of backs for example) the other removes a particular
> >> interaction from the appropriate array of interactions (pull from the
> array
> >> of backs for example). In Mongo, $push and $pull are atomic in that if
> >> someone else calls a $push or $pull to the same idea, or even same
> >> interaction, they do not affect each other.
> >>
> >> So in the context here, we're trying to pull a particular element from
> >> the array of backs.
> >>
> >> Does this at least shed light on the push and pull?
> >> On Dec 11, 2015 2:28 PM, "monotkate" <notifications@github.com> wrote:
> >>
> >>> Nope, this is a backend issue. I'm actually passing it the backing
> >>> object super easily in the code, it's just not in the function call.
> >>>
> >>> I tracked it to the backend as well. Something funky is going on with
> >>> your new interaction setup. the db function for updateOnePullArray is
> >>> getting my back object, but it's not deleting it. This is the same as
> the
> >>> other issue where I just don't know where to troubleshoot it past that
> >>> point.
> >>>
> >>> [image: image]
> >>> <
> https://cloud.githubusercontent.com/assets/11742296/11753430/14bfd3ea-a013-11e5-8693-cf0fc4535b03.png
> >
> >>> [image: image]
> >>> <
> https://cloud.githubusercontent.com/assets/11742296/11753453/28cd6faa-a013-11e5-8044-2980e0fd7035.png
> >
> >>>
> >>> I've never been entirely sure why this one is called updateOnePullArray
> >>> and then updateOnePushArray either, but here's my console log:
> >>> [image: image]
> >>> <
> https://cloud.githubusercontent.com/assets/11742296/11753481/4cc7d76a-a013-11e5-8a15-c3643afc390e.png
> >
> >>>
> >>> —
> >>> Reply to this email directly or view it on GitHub
> >>> <
> https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-164026695
> >
> >>> .
> >>>
> >>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-164040289>
> .
>

Mctalian commented:
Because technically it's not an error, it just finds no results so it does
nothing. So we probably should implement a check to at least notify us of
no results. So in short, it doesn't say there was an error because we've
never run into this as an issue before and I didn't have the foresight to
include an error message for this scenario.

Probably issue worthy if you want to log it for when the backend work is
done.
On Dec 11, 2015 3:34 PM, "monotkate" <notifications@github.com> wrote:

> why doesn't it error then if it doesn't work?
>
> On Fri, Dec 11, 2015 at 3:28 PM, Robert <notifications@github.com> wrote:
>
> > Explanation: the DB looks for an object that matches the exact query,
> since
> > we're passing objects we need to pass the exact object as seen in the DB.
> > It needs to follow the database schema in other words and isInTeam is not
> > in the real schema, just something we use on the client side for building
> > the team up. Those objects aren't guaranteed to have a unique identifier
> > since they roll up under the idea itself, so we need to search for the
> > entire object itself. (i.e. comments have an _id, but updates do not)
> >
> > On Dec 11, 2015 3:17 PM, "Robert Anderson" <r-anderson.9@onu.edu> wrote:
> >
> > > Ah, we need to remove isInTeam field from the back object! Check out
> > > removeInteraction how we remove the author object and also remove
> > isInTeam
> > > and it should work! :)
> > > On Dec 11, 2015 3:16 PM, "Robert Anderson" <r-anderson.9@onu.edu>
> wrote:
> > >
> > >> One adds an interaction to the array of that particular interaction
> > (push
> > >> to the array of backs for example) the other removes a particular
> > >> interaction from the appropriate array of interactions (pull from the
> > array
> > >> of backs for example). In Mongo, $push and $pull are atomic in that if
> > >> someone else calls a $push or $pull to the same idea, or even same
> > >> interaction, they do not affect each other.
> > >>
> > >> So in the context here, we're trying to pull a particular element from
> > >> the array of backs.
> > >>
> > >> Does this at least shed light on the push and pull?
> > >> On Dec 11, 2015 2:28 PM, "monotkate" <notifications@github.com>
> wrote:
> > >>
> > >>> Nope, this is a backend issue. I'm actually passing it the backing
> > >>> object super easily in the code, it's just not in the function call.
> > >>>
> > >>> I tracked it to the backend as well. Something funky is going on with
> > >>> your new interaction setup. the db function for updateOnePullArray is
> > >>> getting my back object, but it's not deleting it. This is the same as
> > the
> > >>> other issue where I just don't know where to troubleshoot it past
> that
> > >>> point.
> > >>>
> > >>> [image: image]
> > >>> <
> >
> https://cloud.githubusercontent.com/assets/11742296/11753430/14bfd3ea-a013-11e5-8693-cf0fc4535b03.png
> > >
> > >>> [image: image]
> > >>> <
> >
> https://cloud.githubusercontent.com/assets/11742296/11753453/28cd6faa-a013-11e5-8044-2980e0fd7035.png
> > >
> > >>>
> > >>> I've never been entirely sure why this one is called
> updateOnePullArray
> > >>> and then updateOnePushArray either, but here's my console log:
> > >>> [image: image]
> > >>> <
> >
> https://cloud.githubusercontent.com/assets/11742296/11753481/4cc7d76a-a013-11e5-8a15-c3643afc390e.png
> > >
> > >>>
> > >>> —
> > >>> Reply to this email directly or view it on GitHub
> > >>> <
> >
> https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-164026695
> > >
> > >>> .
> > >>>
> > >>
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-164040289
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/YashdalfTheGray/flintandsteel/pull/188#issuecomment-164041429>
> .
>
